### DISCUSSION TITLE
Unknown rich dependency op 'python3': Requires:       (python(abi) =  if python3)

### DISCUSSION THREAD

**Discussion by brianjmurrell** [rpm-software-management/mock#1086] [Q&A]
When I try to build https://git.centos.org/rpms/mpich/blob/c8s/f/SPECS/mpich.spec with mock 3.5 in a rocky+epel-8-x86_64 chroot on Fedora 37, I get:

```
Mock Version: 3.5
ENTER ['do_with_status'](['bash', '--login', '-c', '/usr/bin/rpmbuild -bs --noclean --target x86_64 --nodeps /builddir/build/SPECS/mpich-EL_8.spec'], chrootPath='/var/lib/mock/rocky+epel-8-x86_64/root'env={'TERM': 'vt100', 'SHELL': '/bin/bash', 'HOME': '/builddir', 'HOSTNAME': 'mock', 'PATH': '/usr/bin:/bin:/usr/sbin:/sbin', 'PROMPT_COMMAND': 'printf "\\033]0;<mock-chroot>\\007"', 'PS1': '<mock-chroot> \\s-\\v\\$ ', 'LANG': 'C.UTF-8'}shell=Falselogger=<mockbuild.trace_decorator.getLog object at 0x7f30f2bfdf10>timeout=0uid=1101gid=135user='mockbuild'nspawn_args=[]unshare_net=TrueprintOutput=False)
Executing command: ['bash', '--login', '-c', '/usr/bin/rpmbuild -bs --noclean --target x86_64 --nodeps /builddir/build/SPECS/mpich-EL_8.spec'] with env {'TERM': 'vt100', 'SHELL': '/bin/bash', 'HOME': '/builddir', 'HOSTNAME': 'mock', 'PATH': '/usr/bin:/bin:/usr/sbin:/sbin', 'PROMPT_COMMAND': 'printf "\\033]0;<mock-chroot>\\007"', 'PS1': '<mock-chroot> \\s-\\v\\$ ', 'LANG': 'C.UTF-8'} and shell False
sh: /usr/bin/python3.6: No such file or directory
error: line 44: Unknown rich dependency op 'python3': Requires:       (python(abi) =  if python3)
Building target platforms: x86_64
Building for target x86_64
Child return code was: 1
EXCEPTION: [Error('Command failed: \n # bash --login -c /usr/bin/rpmbuild -bs --noclean --target x86_64 --nodeps /builddir/build/SPECS/mpich-EL_8.spec\n', 1)]
Traceback (most recent call last):
  File "/usr/lib/python3.11/site-packages/mockbuild/trace_decorator.py", line 93, in trace
    result = func(*args, **kw)
             ^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/mockbuild/util.py", line 598, in do_with_status
    raise exception.Error("Command failed: \n # %s\n%s" % (command, output), child.returncode)
mockbuild.exception.Error: Command failed: 
 # bash --login -c /usr/bin/rpmbuild -bs --noclean --target x86_64 --nodeps /builddir/build/SPECS/mpich-EL_8.spec
```
in the `root.log`.

This same spec worked fine with mock 2.16.

This same speck also works with mock 3.5 in a rocky+epel-9-x86_64 chroot on Fedora 37, but the important question is why it worked on mock 2.16 and now doesn't work on mock 3.5?

**Comment by brianjmurrell** [comment]
I'm noticing in the above output:

```
sh: /usr/bin/python3.6: No such file or directory
```

That seems like it could be a problem.  What is trying to use `python36` at that stage of mock building?

**Reply by praiskup** [reply]
> sh: /usr/bin/python3.6: No such file or directory

This is typically an [innocent warning](https://github.com/rpm-software-management/rpm/issues/424).

> This same spec worked fine with mock 2.16.

That would be weird.  Are you sure `(python(abi) =  if python3` worked against a RHEL 8 based chroot?

**Reply by brianjmurrell** [reply]
> > sh: /usr/bin/python3.6: No such file or directory
> 
> This is typically an [innocent warning](https://github.com/rpm-software-management/rpm/issues/424).

I'm not so sure it's innocent.  When it happens, my `%{python3_version}` macro is empty (when it shouldn't be).

> That would be weird. Are you sure `(python(abi) = if python3` worked against a RHEL 8 based chroot?

With mock 2.16 yes.  With mock 3.5, no.  Was there perhaps some difference in where that might have been evaluated in mock 2.16 compared to 3.5?  Like perhaps with 2.16 it was done outside of the chroot and in 3.5 it was moved to being done inside?

**Reply by praiskup** [reply]
2.16 is quite old, but from what I remember - source RPM and binary RPMs were always generated *in* the "target" chroot/buildroot.

**Reply by praiskup** [reply]
> This same spec worked fine with mock 2.16.

What do you mean by "worked"?  With older spec file?  If you have the same spec file, can't you test it with 2.16 just to be sure?

**Reply by praiskup** [reply]
Wait, on RHEL 8 there's rpm-4.14.3-26.el8.x86_64 which supports conditional dependencies.

**Comment by praiskup** [comment]
Seems like your spec uses `Requires: (python(abi) = %{python3_version} if python3)`.  This statement to work implies that  `python3-devel` is not only installed at RPM build time, but also when generating source RPM (from spec + sources).  See also the related [rpm issue](https://github.com/rpm-software-management/rpm/issues/424) for more info.

You probably want something like `Requires: (python(abi) = %{?python3_version}%{!?python3_version:666} if python3)`.  This goes innocently at SRPM build time but expands correctly at RPM build time.

Rather then related to a previous version of Mock, this looks that python3 related tooling was removed from the minimal buildroot of rocky+epel-8.

**Comment by brianjmurrell** [comment]
Something more subtle is going on here.  If I add the following to my specfile:
```
%{warn:python3_version: %{python3_version}}
%{warn:debug %{?python3_version}%{!?python3_version:666}}
%{warn:debug %{!?python3_version:123}}
%{warn:debug %{!?nopython3_version:666}}
```

I get the following results:

```
warning: python3_version: 
sh: /usr/bin/python3.6: No such file or directory
warning: debug 
warning: debug 
warning: debug 666
```

So, ultimately the `%{!?python3_version:666}` is not working so `%{python3_version}` must be defined, but empty.
